### PR TITLE
Resolve #368: RPC for data receive and load

### LINKED ISSUE #368: RPC for data receive and load
Use case: drag&drop file(s) from the client, send the blob to Brayns and load the blob (mesh, xyz, ...) accordingly

### PR DESCRIPTION
**Pull request by tribal-tec**

### CODE REVIEW

**Comment by favreau** [comment]
Fantastic PR! I will have a look at it tomorrow.

File: brayns/io/MeshLoader.cpp [hunk 1]
```diff
@@ -333,6 +309,106 @@ void MeshLoader::_createMaterials(Scene& scene, const aiScene* aiScene,
...
+        }
+        if (nonTriangulatedFaces)
+            BRAYNS_DEBUG
+                << "Some faces are not triangulated and have been removed"
+                << std::endl;
+
+        _meshIndex[materialId] += mesh->mNumVertices;
+
+        updateProgress("Post-processing mesh ...",
+                       50 + ((m + 1) * 50.f) / aiScene->mNumMeshes, 100);
```
Comment (rdumusc) [inline brayns/io/MeshLoader.cpp]: looks like the other half of the magic :-) I don't know if we could write a sort of proxy progress class that does the aggregation of the multiple steps.

File: brayns/tasks/UploadBinaryTask.cpp [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+    // pre-check for validity of given params
+    for (size_t i = 0; i < params.size(); ++i)
+    {
+        const auto& param = params[i];
+        if (param.type.empty() || param.size == 0)
+            throw MISSING_PARAMS;
+
+        auto found = std::find_if(supportedTypes.cbegin(),
+                                  supportedTypes.cend(), [&](auto val) {
```
Comment (rdumusc) [inline brayns/tasks/UploadBinaryTask.cpp]: const auto&

File: brayns/tasks/UploadBinaryTask.cpp [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        throw MISSING_PARAMS;
+
+    // pre-check for validity of given params
+    for (size_t i = 0; i < params.size(); ++i)
+    {
+        const auto& param = params[i];
+        if (param.type.empty() || param.size == 0)
+            throw MISSING_PARAMS;
+
+        auto found = std::find_if(supportedTypes.cbegin(),
```
Comment (rdumusc) [inline brayns/tasks/UploadBinaryTask.cpp]: const

File: brayns/tasks/UploadBinaryTask.cpp [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+{
+UploadBinaryTask::UploadBinaryTask(const BinaryParams& params,
+                                   const std::set<std::string>& supportedTypes,
+                                   EnginePtr engine)
+    : _params(params)
+{
+    if (params.empty())
+        throw MISSING_PARAMS;
+
+    // pre-check for validity of given params
```
Comment (rdumusc) [inline brayns/tasks/UploadBinaryTask.cpp]: make it a function, this ctor is too long.

File: brayns/tasks/UploadPathTask.h [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+#include <brayns/common/tasks/Task.h>
+
+namespace brayns
+{
+/**
+ * A task which loads data from a list of paths and adds the loaded data to the
+ * scene.
+ */
+class UploadPathTask : public TaskT<bool>
```
Comment (rdumusc) [inline brayns/tasks/UploadPathTask.h]: I'm realizing now that TaskT is a strange class name, by convention this generally denotes a typename in a template. How about renaming TaskT -> Task and Task -> AbstractTask?

File: brayns/tasks/errors.h [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    BinaryTaskError(const std::string& message, const int code,
+                    const BinaryError& error)
+        : TaskRuntimeError(message, code)
+        , _error(error)
+    {
+    }
+
+    const auto& error() const { return _error; }
+private:
+    const BinaryError _error;
```
Comment (rdumusc) [inline brayns/tasks/errors.h]: could be public

File: brayns/tasks/UploadPathTask.h [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+public:
+    UploadPathTask(std::vector<std::string>&& paths,
+                   const std::set<std::string>& supportedTypes,
+                   EnginePtr engine);
+
+private:
+    std::vector<async::task<void>> _loadTasks;
+};
+
+auto createUploadPathTask(std::vector<std::string>&& paths, const uintptr_t,
```
Comment (rdumusc) [inline brayns/tasks/UploadPathTask.h]: I don't really see the point of this strange method here.

File: plugins/RocketsPlugin/RocketsPlugin.cpp [hunk 8]
```diff
@@ -350,15 +365,146 @@ class RocketsPlugin::Impl
...
+            // transform task error to rockets error response
+            auto errorCallback = [respond](const TaskRuntimeError& error) {
+                respond({Response::Error{error.what(), error.code(),
+                                         error.data()}});
+            };
+
+            try
+            {
+                // transform task result to rockets response
+                auto readyCallback = [respond](R result) {
```
Comment (rdumusc) [inline plugins/RocketsPlugin/RocketsPlugin.cpp]: const R&

File: brayns/common/tasks/Task.h [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+ * successful execution is of type T.
+ *
+ * It accepts any functors of type TaskFunctor will be provided with cancel
+ * support and progress feedback possibility.
+ */
+template <typename T>
+class TaskT : public Task
+{
+public:
+    using Type = async::task<T>;
```
[the remark this answers is not in the document]
Comment (tribal-tec) [reply brayns/common/tasks/Task.h]: Is used in RocketsPlugin:453

File: brayns/common/tasks/TaskFunctor.h [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+#include <async++.h>
+#ifdef __GNUC__
+#pragma GCC diagnostic pop
+#endif
+
+namespace brayns
+{
+/**
+ * A base class for functors that can be used for Tasks. It provides support for
+ * progress supporting and cancellation during execution.
```
[the remark this answers is not in the document]
Comment (tribal-tec) [reply brayns/common/tasks/TaskFunctor.h]: should have been 'reporting'

File: brayns/Brayns.cpp [hunk 11]
```diff
@@ -580,12 +584,13 @@ struct Brayns::Impl : public PluginAPI
                               Vector3f(volumeDimensions) *
                                   volumeElementSpacing);
         }
+        return true;
```
[the remark this answers is not in the document]
Comment (tribal-tec) [reply brayns/Brayns.cpp]: not anymore

**Pull request merged**